dkc/core/rest/file.py

```diff
@@ -83,11 +101,42 @@ def get_serializer_class(self):
             return FileUpdateSerializer
         return FileSerializer
 
+    def _validate_authorized_upload(self, authorization: str, folder: Folder) -> User:
+        try:
+            signed_obj = signing.loads(authorization)
+        except signing.BadSignature:
+            logger = logging.getLogger('django.security.SuspiciousOperation')
+            logger.warning('Authorized upload signature tampering detected.')
+            raise PermissionDenied('Invalid authorization signature.')
+
+        if signed_obj.get('scope') != 'authorized_upload':
+            raise PermissionDenied('Invalid signature scope.')
+
+        try:
+            upload: AuthorizedUpload = AuthorizedUpload.objects.select_related('creator').get(
+                pk=signed_obj['id']
+            )
+        except AuthorizedUpload.DoesNotExist:
+            raise PermissionDenied('This upload authorization has been revoked.')
+
+        if upload.expires < timezone.now():
+            raise PermissionDenied(
+                'This upload authorization has expired. Please request a new link.'
+            )
+        if upload.folder.id != folder.id:
+            raise PermissionDenied('This upload was authorized to a different folder.')
+
+        return upload.creator
+
     def perform_create(self, serializer: FileSerializer):
         folder: Folder = serializer.validated_data['folder']
-        user: User = self.request.user
+        if 'authorization' in self.request.data:
```

Comment:
I think it's unsafe to look at `self.request.data` directly. Given that POST bodies are JSON and auto-decoded, it might not even be a `str` (which is what the parameter type of `_validate_authorized_upload` expects).

We can tack an extra write-only field on `FileSerializer` to obtain this value safely.

Reply:
Also, I'm hoping that maybe a custom DRF field would allow the `signing.loads` to be abstracted within the serializer, so we're always guaranteed to receive a valid signed value as output.